File: ticktock-compiler/src/main/kotlin/dev/zacsweers/ticktock/compiler/JavaWriter.java
package dev.zacsweers.ticktock.compiler;

import static javax.lang.model.element.Modifier.FINAL;
import static javax.lang.model.element.Modifier.STATIC;

import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.FieldSpec;
import com.squareup.javapoet.JavaFile;
import com.squareup.javapoet.ParameterizedTypeName;
import com.squareup.javapoet.TypeName;
import com.squareup.javapoet.TypeSpec;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.Set;

final class JavaWriter implements RulesWriter {

  private final Path outputDir;

  JavaWriter(Path outputDir) {
    this.outputDir = outputDir;
  }

  @Override public void writeZoneIds(String packageName, String version, Set<String> zoneIds)
      throws IOException {
    TypeSpec typeSpec = TypeSpec.classBuilder("LazyZoneRules")
        .addModifiers(FINAL)
        .addField(version(version))
        .addField(regionId(zoneIds))
        .build();

    if (!Files.exists(outputDir)) {
      Files.createDirectories(outputDir);
    }
    JavaFile.builder(packageName, typeSpec)
        .build()
        .writeTo(outputDir);
  }

  private FieldSpec version(String version) {
    return FieldSpec.builder(String.class, "VERSION", STATIC, FINAL)
        .initializer("$S", version)
        .build();
  }

  private FieldSpec regionId(Set<String> allRegionIds) {
    CodeBlock.Builder builder = CodeBlock.builder()
        .add("$T.asList(\n$>$>", Arrays.class);
    Iterator<String> iterator = allRegionIds.iterator();
    while (iterator.hasNext()) {
      builder.add("$S", iterator.next());
      if (iterator.hasNext()) {
        builder.add(",\n");
      }
    }
    builder.add("$<$<)");
    TypeName listType = ParameterizedTypeName.get(List.class, String.class);
    return FieldSpec.builder(listType, "REGION_IDS", STATIC, FINAL)
        .initializer(builder.build())
        .build();
  }
}

File: ticktock-compiler/src/main/kotlin/dev/zacsweers/ticktock/compiler/JavaWriter.kt
/*
 * Copyright (C) 2020 Zac Sweers & Gabriel Ittner
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dev.zacsweers.ticktock.compiler

import com.squareup.javapoet.CodeBlock
import com.squareup.javapoet.FieldSpec
import com.squareup.javapoet.JavaFile
import com.squareup.javapoet.MethodSpec
import com.squareup.javapoet.ParameterizedTypeName
import com.squareup.javapoet.TypeName
import com.squareup.javapoet.TypeSpec
import dev.zacsweers.ticktock.runtime.ZoneIdsProvider
import java.nio.file.Files
import java.nio.file.Path
import java.util.Arrays
import javax.lang.model.element.Modifier.FINAL
import javax.lang.model.element.Modifier.PRIVATE
import javax.lang.model.element.Modifier.PUBLIC
import javax.lang.model.element.Modifier.STATIC

internal class JavaWriter(
  private val outputDir: Path
) : RulesWriter {

  override fun writeZoneIds(packageName: String, version: String, zoneIds: Set<String>) {
    val versionField = version(version)
    val zoneIdsField = regionId(zoneIds)
    val typeSpec = TypeSpec.classBuilder("GeneratedZoneIdsProvider")
        .addModifiers(FINAL)
        .addSuperinterface(ZoneIdsProvider::class.java)
        .addField(versionField)
        .addMethod(MethodSpec.methodBuilder("getVersionId")
            .addAnnotation(Override::class.java)
            .addModifiers(PUBLIC)
            .returns(versionField.type)
            .addStatement("return \$N", versionField)
            .build())
        .addField(zoneIdsField)
        .addMethod(MethodSpec.methodBuilder("getZoneIds")
            .addAnnotation(Override::class.java)
            .addModifiers(PUBLIC)
            .returns(zoneIdsField.type)
            .addStatement("return \$N", zoneIdsField)
            .build())
        .build()

    Files.createDirectories(outputDir)
    JavaFile.builder(packageName, typeSpec)
        .build()
        .writeTo(outputDir)
  }

  private fun version(version: String): FieldSpec {
    return FieldSpec.builder(String::class.java, "VERSION_ID", PRIVATE, STATIC, FINAL)
        .initializer("\$S", version)
        .build()
  }

  private fun regionId(allRegionIds: Set<String>): FieldSpec {
    val blocks = allRegionIds.map { CodeBlock.of("\$S", it) }
    val joinedBlocks = CodeBlock.join(blocks, ",\n")
    val initializer = CodeBlock.builder()
        .add("\$T.asList(\n$>$>", Arrays::class.java)
        .add(joinedBlocks)
        .add("$<$<)")
        .build()

    val listType: TypeName = ParameterizedTypeName.get(List::class.java, String::class.java)
    return FieldSpec.builder(listType, "ZONE_IDS", PRIVATE, STATIC, FINAL)
        .initializer(initializer)
        .build()
  }
}
